src/Swift.Runtime/src/Metadata/TypeMetadata.cs

```diff
@@ -0,0 +1,199 @@
+    /// The metadata represents a struct
+    /// </summary>
+    Struct = 0 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an enum
+    /// </summary>
+    Enum = 1 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an optional type
+    /// </summary>
+    Optional = 2 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an non-swift class
+    /// </summary>
+    ForeignClass = 3 | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an opaque type
+    /// </summary>
+    Opaque = 0 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a tuple
+    /// </summary>
+    Tuple = 1 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a closure/function
+    /// </summary>
+    Function = 2 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a protocol
+    /// </summary>
+    Protocol = 3 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a type of a TypeMetadata type
+    /// </summary>
+    Metatype = 4 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents an Objective C wrapper
+    /// </summary>
+    ObjCClassWrapper = 5 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a type of an existential container
+    /// </summary>
+    ExistentialMetatype = 6 | TypeMetadataFlags.MetadataKindIsRuntimePrivate | TypeMetadataFlags.MetadataKindIsNonHeap,
+    /// <summary>
+    /// The metadata represents a heap local variable
+    /// </summary>
+    HeapLocalVariable = 0 | TypeMetadataFlags.MetadataKindIsNonType,
+    /// <summary>
+    /// The metadata represents a generic heap local variable
+    /// </summary>
+    HeapGenericLocalVariable = 0 | TypeMetadataFlags.MetadataKindIsNonType | TypeMetadataFlags.MetadataKindIsRuntimePrivate,
+    /// <summary>
+    /// The metadata represents an error
+    /// </summary>
+    ErrorObject = 1 | TypeMetadataFlags.MetadataKindIsNonType | TypeMetadataFlags.MetadataKindIsRuntimePrivate,
+    // Swift source code says that for fixed values, this will never exceed 0x7ff,
+    // but all class types will be 0x800 and above
+    /// <summary>
+    /// The metadata represents a class
+    /// </summary>
+    Class = 0x800 
+}
+
+
+/// <summary>
+/// Represents the type metadata for a Swift type
+/// </summary>
+public readonly struct TypeMetadata : IEquatable<TypeMetadata> {
+    readonly IntPtr handle;
```

Comment:
Will this become `SwiftHandle` in the future?

Reply:
That's a great question - since this is a private member and we should never (1) need to access this outside of the TypeMetadata (and I don't think we ever would) and (2) get this type from any other place that as an implicit argument passed to us from Swift or as a return from an explicit call, it's fine to leave it as an IntPtr.